### PR TITLE
Reducer LOB usage in Batch2 and Search

### PR DESCRIPTION
**Pull request by jamesagnew**

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/Search.java [hunk 1]
```diff
@@ -169,10 +179,14 @@ public class Search implements ICachedSearchDetails, Serializable {
 	@Column(name = "OPTLOCK_VERSION", nullable = true)
 	private Integer myVersion;
 
-	@Lob
+	@Lob // TODO: VC column added in 7.2.0 - Remove non-VC column later
 	@Column(name = "SEARCH_PARAM_MAP", nullable = true)
 	private byte[] mySearchParameterMap;
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/Search.java:184]: just a suggestion to triple check all accessors and make sure the only access method is that public `getSearchQueryString()`

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/Search.java [hunk 2]
```diff
@@ -169,10 +179,14 @@ public class Search implements ICachedSearchDetails, Serializable {
...
 
+	@Column(name = "SEARCH_PARAM_MAP_BIN", nullable = true)
+	@Type(type = JpaConstants.ORG_HIBERNATE_TYPE_BINARY_TYPE)
+	private byte[] mySearchParameterMapBin;
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/Search.java:188]: is this not also a byte[] like the old one? I would have assumed you would use a string here too. Or is the @Type sufficient to make it not use a LOB?

File: hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/ColumnTypeEnum.java [hunk 3]
```diff
@@ -31,5 +31,8 @@ public enum ColumnTypeEnum {
 	BLOB,
 	CLOB,
 	DOUBLE,
-	TEXT;
+	/** Long inline text */
+	TEXT,
+	/** Long inline binary */
+	BINARY;
```
Comment (tadgh) [inline hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/ColumnTypeEnum.java:37]: ah i see, inline binary

File: hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/ColumnTypeToDriverTypeToSqlType.java [hunk 4]
```diff
@@ -130,6 +130,14 @@ private ColumnTypeToDriverTypeToSqlType() {}
...
 		setColumnType(ColumnTypeEnum.TEXT, DriverTypeEnum.POSTGRES_9_4, "text");
 		setColumnType(ColumnTypeEnum.TEXT, DriverTypeEnum.MSSQL_2012, "varchar(MAX)");
+
+		setColumnType(ColumnTypeEnum.BINARY, DriverTypeEnum.H2_EMBEDDED, "binary(2147483647)");
+		setColumnType(ColumnTypeEnum.BINARY, DriverTypeEnum.DERBY_EMBEDDED, "varchar(2147483647) for bit data");
+		setColumnType(ColumnTypeEnum.BINARY, DriverTypeEnum.MARIADB_10_1, "longblob");
+		setColumnType(ColumnTypeEnum.BINARY, DriverTypeEnum.MYSQL_5_7, "longblob");
+		setColumnType(ColumnTypeEnum.BINARY, DriverTypeEnum.ORACLE_12C, "blob");
+		setColumnType(ColumnTypeEnum.BINARY, DriverTypeEnum.POSTGRES_9_4, "bytea");
+		setColumnType(ColumnTypeEnum.BINARY, DriverTypeEnum.MSSQL_2012, "varbinary(MAX)");
```
Comment (tadgh) [inline hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/ColumnTypeToDriverTypeToSqlType.java:140]: We will need to have QA spin up all these DBs and make sure this works, I suppose. I imagine you have tested 1-2 of them?

**Review by tadgh — APPROVED** [review]
LGTM pending a few questions, suggerstions.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/Search.java (hunk 2 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/Search.java:188]: Yeah, that `@Type` causes it to use a `bytea` datatype on postgres, which is an inline (ie non-LOB) byte array. It also uses equivalent datatypes on other databases.

File: hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/ColumnTypeToDriverTypeToSqlType.java (hunk 4 above)
Comment (jamesagnew) [reply hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/ColumnTypeToDriverTypeToSqlType.java:140]: Yup - At least on Postgres. Still working on a unit test that verifies this automatically too...

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/Search.java (hunk 1 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/Search.java:185]: Done

**Pull request merged**